File: structural/typescript/composite_structured/composites.ts
import { FileSystemComponent, Employee, Graphic } from './components';

export class Directory extends FileSystemComponent {
  private name: string;
  private children: FileSystemComponent[] = [];
  
  constructor(name: string) { super(); this.name = name; }
  
  add(component: FileSystemComponent): void { this.children.push(component); }
  remove(component: FileSystemComponent): void {
    const index = this.children.indexOf(component);
    if (index > -1) { this.children.splice(index, 1); }
  }
  getChild(index: number): FileSystemComponent | null {
    if (index >= 0 && index < this.children.length) { return this.children[index]; }
    return null;
  }
  display(indent: string = ''): void {
    console.log(`\${indent}Directory: \${this.name} (\${this.getSize()} bytes)`);
    for (const child of this.children) { child.display(indent + '  '); }
  }
  getSize(): number {
    let totalSize = 0;
    for (const child of this.children) { totalSize += child.getSize(); }
    return totalSize;
  }
}

export class Manager extends Employee {
  private name: string;
  private title: string;
  private subordinates: Employee[] = [];
  
  constructor(name: string, title: string) { super(); this.name = name; this.title = title; }
  
  add(employee: Employee): void { this.subordinates.push(employee); }
  remove(employee: Employee): void {
    const index = this.subordinates.indexOf(employee);
    if (index > -1) { this.subordinates.splice(index, 1); }
  }
  print(indent: string = ''): void {
    console.log(`\${indent}Manager: \${this.name} - \${this.title}`);
    for (const subordinate of this.subordinates) { subordinate.print(indent + '  '); }
  }
}

export class GraphicGroup extends Graphic {
  private name: string;
  private graphics: Graphic[] = [];
  
  constructor(name: string) { super(); this.name = name; }
  
  add(graphic: Graphic): void { this.graphics.push(graphic); }
  remove(graphic: Graphic): void {
    const index = this.graphics.indexOf(graphic);
    if (index > -1) { this.graphics.splice(index, 1); }
  }
  draw(indent: string = ''): void {
    console.log(`\${indent}Group: \${this.name}`);
    for (const graphic of this.graphics) { graphic.draw(indent + '  '); }
  }
}

File: structural/dart/composite_structured/composites.dart
import 'components.dart';

class Directory extends FileSystemComponent {
  final String name;
  final List<FileSystemComponent> _children = [];
  Directory(this.name);
  
  @override
  void add(FileSystemComponent component) { _children.add(component); }
  @override
  void remove(FileSystemComponent component) { _children.remove(component); }
  @override
  FileSystemComponent? getChild(int index) {
    if (index >= 0 && index < _children.length) { return _children[index]; }
    return null;
  }
  
  @override
  void display([String indent = '']) {
    print('\$indent Directory: \$name (\${getSize()} bytes)');
    for (var child in _children) { child.display(indent + '  '); }
  }
  
  @override
  int getSize() {
    int totalSize = 0;
    for (var child in _children) { totalSize += child.getSize(); }
    return totalSize;
  }
}

class Manager extends Employee {
  final String name;
  final String title;
  final List<Employee> _subordinates = [];
  Manager(this.name, this.title);
  
  @override
  void add(Employee employee) { _subordinates.add(employee); }
  @override
  void remove(Employee employee) { _subordinates.remove(employee); }
  
  @override
  void print([String indent = '']) {
    print('\$indent Manager: \$name - \$title');
    for (var subordinate in _subordinates) { subordinate.print(indent + '  '); }
  }
}

class GraphicGroup extends Graphic {
  final String name;
  final List<Graphic> _graphics = [];
  GraphicGroup(this.name);
  
  void add(Graphic graphic) { _graphics.add(graphic); }
  void remove(Graphic graphic) { _graphics.remove(graphic); }
  
  @override
  void draw([String indent = '']) {
    print('\$indent Group: \$name');
    for (var graphic in _graphics) { graphic.draw(indent + '  '); }
  }
}
